src/state_definitions.jl

```diff
@@ -16,7 +15,14 @@ randstate(b) = randstate(ComplexF64, b)
 
 Returns a Haar random pure state of dimension `length(b)` by applying a Haar random unitary to a fixed pure state.
 """
-randstate_haar(b::Basis) = Ket(b, rand(Haar(2), length(b), 2)[:,1])
+function randstate_haar(b::Basis)
+    dim = length(b)
+    mat = rand(ComplexF64, dim, dim)
+    q, r = qr(mat)
+    d = Diagonal([r[i, i] / abs(r[i, i]) for i in 1:dim])
+    data = q * d
+    return Ket(b, data[:,1])
```

Comment:
Isn't this just the first column of q up to some random global phase? We can probably skip line 22 and 23 if that is the case, making this even faster.